File: sample/src/main/java/net/mm2d/upnp/sample/ServiceNode.java
/*
 * Copyright(C) 2016 大前良介(OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.sample;

import net.mm2d.upnp.Action;
import net.mm2d.upnp.Service;

import java.awt.Component;
import java.awt.Desktop;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.IOException;
import java.net.URISyntaxException;
import java.util.List;

import javax.swing.JFrame;
import javax.swing.JMenuItem;
import javax.swing.JPopupMenu;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介(OHMAE Ryosuke)</a>
 */
public class ServiceNode extends UpnpNode {
    private boolean mSubscribing;

    public ServiceNode(final Service service) {
        super(service);
        final List<Action> actions = service.getActionList();
        for (final Action action : actions) {
            add(new ActionNode(action));
        }
        add(new StateVariableListNode(service.getStateVariableList()));
    }

    @Override
    public Service getUserObject() {
        return (Service) super.getUserObject();
    }

    @Override
    public String getDetailText() {
        final Service service = getUserObject();
        return "ServiceType: " +
                service.getServiceType() +
                '\n' +
                "ServiceId: " +
                service.getServiceId() +
                '\n' +
                "ScpdUrl: " +
                service.getScpdUrl() +
                '\n' +
                "ControlUrl: " +
                service.getControlUrl() +
                '\n' +
                "EventSubUrl: " +
                service.getEventSubUrl();
    }

    @Override
    public String getDetailXml() {
        return formatXml(getUserObject().getDescription());
    }

    @Override
    public String toString() {
        return getUserObject().getServiceType();
    }

    @Override
    public void showContextMenu(
            final JFrame frame,
            final Component invoker,
            final int x,
            final int y) {
        final JPopupMenu menu = new JPopupMenu();
        final JMenuItem open = new JMenuItem("Open Service Description");
        open.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(final ActionEvent e) {
                final Service service = getUserObject();
                try {
                    Desktop.getDesktop().browse(service.getAbsoluteUrl(service.getScpdUrl()).toURI());
                } catch (IOException | URISyntaxException e1) {
                    e1.printStackTrace();
                }
            }
        });
        menu.add(open);
        if (mSubscribing) {
            final JMenuItem unsubscribe = new JMenuItem("Unsubscribe");
            unsubscribe.addActionListener(new ActionListener() {
                @Override
                public void actionPerformed(final ActionEvent e) {
                    try {
                        getUserObject().unsubscribe();
                        mSubscribing = false;
                    } catch (final IOException e1) {
                        e1.printStackTrace();
                    }
                }
            });
            menu.add(unsubscribe);
        } else {
            final JMenuItem subscribe = new JMenuItem("Subscribe");
            subscribe.addActionListener(new ActionListener() {
                @Override
                public void actionPerformed(final ActionEvent e) {
                    try {
                        getUserObject().subscribe(true);
                        mSubscribing = true;
                    } catch (final IOException e1) {
                        e1.printStackTrace();
                    }
                }
            });
            menu.add(subscribe);
        }
        menu.show(invoker, x, y);
    }

    public boolean isSubscribing() {
        return mSubscribing;
    }
}

File: sample/src/main/java/net/mm2d/upnp/sample/ServiceNode.kt
/*
 * Copyright (c) 2016 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.sample

import kotlinx.coroutines.runBlocking
import net.mm2d.upnp.Http
import net.mm2d.upnp.Service
import java.awt.Component
import java.awt.Desktop
import java.io.IOException
import java.net.URISyntaxException
import javax.swing.JFrame
import javax.swing.JMenuItem
import javax.swing.JPopupMenu

class ServiceNode(service: Service) : UpnpNode(service) {
    var isSubscribing: Boolean = false
        private set

    override fun getDetailXml(): String = formatXml(getUserObject().description)

    init {
        val actions = service.actionList
        for (action in actions) {
            add(ActionNode(action))
        }
        add(StateVariableListNode(service.stateVariableList))
    }

    override fun getUserObject(): Service = super.getUserObject() as Service
    override fun formatDescription(): String = Formatter.format(getUserObject())
    override fun toString(): String = getUserObject().serviceType

    override fun showContextMenu(frame: JFrame, invoker: Component, x: Int, y: Int) {
        val menu = JPopupMenu()
        menu.add(JMenuItem("Open Service Description").also {
            it.addActionListener { openServerDescription() }
        })
        if (isSubscribing) {
            menu.add(JMenuItem("Unsubscribe").also {
                it.addActionListener { unsubscribe() }
            })
        } else {
            menu.add(JMenuItem("Subscribe").also {
                it.addActionListener { subscribe() }
            })
        }
        menu.show(invoker, x, y)
    }

    private fun openServerDescription() {
        val service = getUserObject()
        val device = service.device
        try {
            val uri = Http.makeAbsoluteUrl(device.baseUrl, service.scpdUrl, device.scopeId).toURI()
            Desktop.getDesktop().browse(uri)
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: URISyntaxException) {
            e.printStackTrace()
        }
    }

    private fun subscribe() {
        isSubscribing = runBlocking { getUserObject().subscribe(true) }
    }

    private fun unsubscribe() {
        isSubscribing = !runBlocking { getUserObject().unsubscribe() }
    }
}
